README.md

```diff
@@ -1,486 +1,679 @@
 
-@localMeta class C {
-  @localMeta method() { }
+```mjs
+@annotate({x: "y"}) @annotate({v: "w"}) class C {
+  @annotate({a: "b"}) method() { }
+  @annotate({c: "d"}) field;
 }
 
-Reflect.getMetadata(C, "key");                      // "value"
-Reflect.getMetadata(C.prototype, "key", "method");  // "value"
+C[Symbol.metadata].class.x                     // "y"
+C[Symbol.metadata].class.v                     // "w"
+C[Symbol.metadata].prototype.methods.method.a  // "a"
+C[Symbol.metadata].instance.fields.field.c     // "d"
 ```
 
-`@metadata` can also be defined in terms of `@register`. When `@register` is used with a public field, method or accessor, it is called with the second argument being the property key of that class element.
+**NOTE:** The exact format of the annotations object is not very well-thought-out and could use more refinement. The main thing I'd like to illustrate here is, it's just an object, with no particular support library to read or write it, and it's automatically created by the system.
 
-```mjs
-// metadata.mjs
-import "reflect-metadata";
+This decorator `@annotate` could be implemented as follows:
 
-export decorator @metadata(key, value) {
-  @register((target, prop) => Reflect.defineMetadata(key, value, target, prop))
+```js
+function annotate(metadata) {
+  return (_, context) => {
+    context.metadata = metadata;
+    return _;
+  }
 }
 ```
 
-### `@frozen`
+Each time a decorator is called, it is passed a new context object, and after each decorator returns, the context object's `metadata` property is read, and if it's not undefined, it's included in the `[Symbol.metadata]` for that class element.
 
-The `@frozen` decorator freezes the constructor and its `prototype` so that they cannot be mutated after the class is defined. It leaves instances mutable. Example usage:
-
-```mjs
-import { @frozen } from "./frozen.mjs";
-
-@frozen
-class MyClass {
-  method() { }
+<!--
+The desugarings in this article usually omit it, but metadata is desugared as such:
+class C {
+  method() {  }
 }
+let context = { kind: "method", name: "method", isStatic: false };
+C.prototype.method = annotate(C.prototype.method, context);
+C[Symbol.metadata].prototype.methods.method = {...{a: "b"}, ...(context[Symbol.metadata] ?? {})};
+-->
 
-MyClass.method = () => {};            // TypeError to add a method
-MyClass.prototype.method = () => {};  // TypeError to overwrite a method
-MyClass.prototype.method.foo = 1;     // TypeError to mutate a method
-```
+Note that, since metadata is held on the class, not on the method, the metadata is not visible to earlier decorators. Metadata on classes is added to the constructor after all class decorators have run so that they are not lost by later wrapping.
```

Comment:
I'm confused; I would expect metadata to be set on the thing being decorated - ie, if a method decorator, on a method; if a class decorator, on a class. Are you saying there's no way to add metadata to a class method?

Reply:
Static and prototype methods can both have metadata. What still needs to be worked out is, what path/paths can you take to get to the metadata from a class/method--the exact object layout. The restriction here is that decorators can't see each other's metadata while they are running.